### ISSUE TITLE
`create-domains` command

### ISSUE THREAD

**Issue by papakai** [StackExchange/dnscontrol#387]
Hi,

opening this new issue to get things clarified separately to not pollute our PR too much. In addition, we do not see it being a necessary part of the PR.

`dnscontrol create-domains` is the only thing I was able to find in the whole documentation, found at [R53](https://github.com/StackExchange/dnscontrol/blob/master/docs/_providers/route53.md) provider info page. I hope I have not missed something. 

*Now coming to my questions:*
* From what I could figure out up to now is that this command finally calls `EnsureDomainExists`. Am I right?
* How does `dnscontrol create-domains` exactly work? Are there further command line arguments? Does it use customer dnsconfig.js to check which domains are not yet on his account? So - where does it take the list of domains from?
* as domain registration is bound to costs: regular price if domain is not taken, aftermarket price if domain is available on some aftermarket platform etc. Showing the availability and the appropriate pricing is a must before adding it to customers account. In case I am right with my 1st question, this could be covered in EnsureDomainExists by adding an order to customer's shopping card in our system and to display details like price and availability details in terminal and asking customer for confirmation... something like a [Y/n] wizard. In case customer proceeds with 'Y', we execute that order. Does that make sense/is that possible?

Thanks so much!
Kai

**Comment by captncraig** [comment]
I don't think the intention with that command was ever to be creating orders or spending money to register domains. We made this for things like route 53, cloudflare, digitial ocean and such that are really **just** dns hosts. They don't care where your domains are registered, so you can create whatever domains you like and they will happily run nameservers for them, whether or not you own them.

Since hexonet doesn't really offer standalone dns hosting, I'd probably recommend not implementing that. 

The command essentially tells the driver "make sure all of these domains are present on the account so we can start managing records for them". The implementation it up to the driver but is generally:

- Get list of domains currently on the account.
- Find any domains in the dns config that are not present.
- Issue a "create" call for each one to make it ready for use.

**Comment by papakai** [comment]
ok thanks for this deeper going explanation.

> Since hexonet doesn't really offer standalone dns hosting

I am not sure if you're right in that point. Anyone could create an account in our system and create a dnszone for a domain not registered through our system and not being in our systems.

So finally this is again some misleading wording imo. as `create-domains` stands more for register domains as create non-existing dnszones (on the appropriate customer account).
Also `EnsureDomainExists` reflects this again. Mouse over tooltip on create domains hardens this misunderstanding.
For sure hard or impossible to review method or command nomenclature, but I highly recommand the documentation parts to be reviewed.

A lot of providers (and I include our module too), check first the existance of a dnszone in their systems and create it if not existing (happens in GetDomainCorrections). This makes that create-domains command then superfluous imo. Should that check not be part of GetDomainCorrections to enforce people to use EnsureDomainExists? I am somehow wondering where the benefit of the command is.

**Comment by papakai** [comment]
> In order to do a valid preview, the zone needs to exist, but it would not seem right to have preview make a real change like creating a zone.

Sounds correct. Especially in our system it leads to costs in case you create a dnszone for a domain that is not being hosted by us, thus a so-called "external" dnszone.
It would be easy to correct that in case the used command is accessible within GetDomainCorrections. That would allow to check if one is in preview mode or in push mode.

I know this is just an idea and it sounds more hacky and against your strict functionality path that I in general like. The question is just - do we follow the way that other providers have implemented (to create the zone in GetDomainCorrections) or to follow your way.
I mean: if we follow your way - people may claim about our module working not working as good as the one of other providers. So this might lead to the impression we are not as good.
If we do not follow your way it might lead to unwanted costs as of creating zones just on preview.

Sounds like this little chaos needs a solution. I personally agree that it would be better to follow your path, but.. :)

Again my question for clarification: is it just to execute `dnscontrol create-domains`? nothing else? no additional parameters? This is undocumented, would be great to get more information on that.

Thanks so much!

**Comment by papakai** [comment]
I rewrote our code accordingly, but Integration Tests were failing now in case the dnszone does not exist.
I deleted it for testing purposes.

Maybe this information could also get clarified here:
https://github.com/StackExchange/dnscontrol/blob/master/docs/writing-providers.md#step-7-integration-test
Instead of 'real' domain, it might be better to use wording 'existing' domain. or even better 'existing' dnszone. This wording was always confusing for me when working with dnscontrol. speaking of a domain where you basically speak about a dnszone.
I've now again created the zone to have the tests running through.

Kai
